sdk/java-sdk-protobuf-testkit/src/main/scala/kalix/javasdk/testkit/impl/EventingTestKitImpl.scala

```diff
@@ -0,0 +1,260 @@
+  }
+
+  final class ServiceImpl extends EventingTestKitService {
+    override def emitSingle(in: EmitSingleCommand): Future[EmitSingleResult] = {
+      log.info("emitSingle request: [{}]", in)
+
+      in.destination.foreach(dest => {
+        val probe = eventDestinationProbe.computeIfAbsent(dest, _ => TestProbe())
+        probe.ref ! in
+      })
+
+      if (in.destination.isEmpty) {
+        log.warn("Received a message without destination, ignoring. {}", in)
+      }
+      Future.successful(EmitSingleResult())
+    }
+
+    override def runSource(in: Source[RunSourceCommand, NotUsed]): Source[SourceElem, NotUsed] = {
+      log.info("runSource request started: {}", in)
+      val runningSourcePromise = Promise[RunningSourceProbe]()
+
+      in.watchTermination() { (_, fDone) =>
+        fDone.onComplete {
+          case Success(_)  => log.debug("runSource in completed")
+          case Failure(ex) => log.error("runSource in failed", ex)
+        }
+      }.runWith(Sink.fold(None: Option[RunningSourceProbe]) {
+        case (
+              None,
+              RunSourceCommand(
+                RunSourceCommand.Command.Create(RunSourceCreate(serviceName, Some(eventSource), _)),
+                _)) =>
+          // proxy triggers probe creation through the create command
+          log.debug(
+            "runSource request got initial create command for service name [{}], source: [{}]",
+            serviceName,
+            eventSource)
+          val (queue, source) = Source.queue[SourceElem](10).preMaterialize()
+          val runningSourceProbe = RunningSourceProbe(serviceName, eventSource)(queue, source)
+          eventSourceProbe.computeIfAbsent(eventSource, _ => TestProbe()).ref ! runningSourceProbe
+          runningSourcePromise.success(runningSourceProbe)
+          Some(runningSourceProbe)
+        case (s @ Some(runningSourceProbe), RunSourceCommand(RunSourceCommand.Command.Ack(sourceAck), _)) =>
+          log.debug("runSource request got ack [{}]", sourceAck)
+          s
+        case wat => throw new MatchError(s"Unexpected fold input: $wat")
+      })
+
+      Source
+        .futureSource(runningSourcePromise.future.map { runningSourceProbe =>
+          runningSourceProbe.outSource
+        })
+        .watchTermination() { (_, fDone) =>
+          fDone.onComplete {
+            case Success(_)  => log.debug("runSource out completed")
+            case Failure(ex) => log.error("runSource out failed", ex)
+          }
+        }
+        .mapMaterializedValue(_ => NotUsed)
+    }
+
+    override def eventStreamOut(in: Source[EventStreamOutCommand, NotUsed]): Source[EventStreamOutResult, NotUsed] = ???
+  }
+}
+
+final class TopicImpl(private val destinationProbe: TestProbe, private val sourceProbe: TestProbe) extends Topic {
+
+  private lazy val brokerProbe = sourceProbe.expectMsgType[RunningSourceProbe]
+  private val log = LoggerFactory.getLogger(classOf[TopicImpl])
+
+  @ApiMayChange
```

Comment:
No need for @ApiMayChange in internals `impl`, we can change anything here (but top types should be marked internal api)